File: model/fmRDSDesign.py
#
# Comp Eng 3DY4 (Computer Systems Integration Project)
#
# Copyright by Nicola Nicolici
# Department of Electrical and Computer Engineering
# McMaster University
# Ontario, Canada
#

import matplotlib.pyplot as plt
from scipy.io import wavfile
from scipy import signal
import numpy as np
import math

# use fmDemodArctan and fmPlotPSD
from fmSupportLib import *
#from fmMonoBlock import *
# for take-home add your functions

rf_Fs = 2.4e6
rf_Fc = 100e3
rf_taps = 151
rf_decim = 10

audio_Fs = 48e3
audio_decim = 5
audio_taps = 151
audio_Fc = 3e3

# RDS VARIABLES=================================================================
# GCD = 125, U = 2375*17/125, D = 240000*U/(2375*17)
sps = 17
rds_sr = sps * 2375
demod_decim = 1920
demod_exp = 323

# add other settings for audio, like filter taps, ...

if __name__ == "__main__":

	# read the raw IQ data from the recorded file
	# IQ data is assumed to be in 8-bits unsigned (and interleaved)
	in_fname = "../data/samples0_2400.raw"
	raw_data = np.fromfile(in_fname, dtype='uint8')
	print("Read raw RF data from \"" + in_fname + "\" in unsigned 8-bit format")
	# IQ data is normalized between -1 and +1 in 32-bit float format
	iq_data = (np.float32(raw_data) - 128.0)/128.0
	print("Reformatted raw RF data to 32-bit float format (" + str(iq_data.size * iq_data.itemsize) + " bytes)")

	# set up the subfigures for plotting
	subfig_height = np.array([0.8, 2, 2, 1.6]) # relative heights of the subfigures
	plt.rc('figure', figsize=(10, 10))	# the size of the entire figure
	fig, (ax0, ax1, ax3, ax2) = plt.subplots(nrows=4, gridspec_kw={'height_ratios': subfig_height})
	fig.subplots_adjust(hspace = .6)

	# FILTER COEFFICIENTS
	rf_coeff = signal.firwin(rf_taps, rf_Fc/(rf_Fs/2), window=('hann'))
	carrier_coeff = myBandPass(113.5e3, 114.5e3, rf_Fs/rf_decim, audio_taps) #For RDS Carrier Recovery
	channel_coeff = myBandPass(54e3, 60e3, rf_Fs/rf_decim, audio_taps) #For RDS Channel Extraction
	rds_demod_coeff = myLowPass(3e3, (rf_Fs/rf_decim)*demod_exp, audio_taps*demod_exp)	# For RDS demodulation
	rrc_coeff = rrc(rds_sr, audio_taps) # for RDS demodulation

	# select a block_size that is a multiple of KB
	# and a multiple of decimation factors
	block_size = 38400*2
	block_count = 0

	# STATE SAVING IQ
	state_i_lpf_100k = np.zeros(rf_taps-1)
	state_q_lpf_100k = np.zeros(rf_taps-1)
	state_phase = 0

	dummy_state = np.array([0,0])						#For RF Demondulation

	# FILTERED DATA
	channel_filt = np.zeros(shape=block_size//2 - 1)
	carrier_filt = np.zeros(shape=block_size//2 - 1)
	channel_Delay = np.zeros(shape=block_size//2 - 1)
	demod_filt = np.zeros(shape=block_size//2 - 1)

	# STATE SAVING
	carrier_block = np.zeros(shape=len(carrier_coeff))	#For Carrier Recovery
	channel_block = np.zeros(shape=len(channel_coeff))	#For Channel Extraction
	delay_block = np.zeros(shape=((audio_taps-1)//2))		#For All pass filter
	rds_demod_blockI =np.zeros(shape=len(rds_demod_coeff))	#For Demodulation Resampler
	rds_demod_blockQ =np.zeros(shape=len(rds_demod_coeff))	#For Demodulation Resampler
	rds_rrc_blockI = np.zeros(shape=len((rrc_coeff)))			#For RRC convolution
	rds_rrc_blockQ = np.zeros(shape=len((rrc_coeff)))			#For RRC convolution

	pll_block = np.array([0.0, 0.0, 1.0, 0.0, 1.0, 1.0, 0.0])
	prevCarrier = np.zeros(shape=5121)
	prevCarrier = np.zeros(shape=5120)

	cdr_stateI = [0, 0, 0]
	initialI = -1

	cdr_stateQ = [0, 0, 0]
	initialQ = -1

	decode_init = -1

	# audio buffer that stores all the audio blocks
	audio_data = np.array([]) # used to concatenate filtered blocks (audio data)

	while (block_count+1)*block_size < len(iq_data):
	#RF FRONT END==================================================
		# if you wish to have shorter runtimes while troubleshooting
		# you can control the above loop exit condition as you see fit
		print('Processing block ' + str(block_count))

		# filter to extract the FM channel (I samples are even, Q samples are odd)
		i_filt, state_i_lpf_100k = signal.lfilter(rf_coeff, 1.0, \
				iq_data[(block_count)*block_size:(block_count+1)*block_size:2],
				zi=state_i_lpf_100k)
		q_filt, state_q_lpf_100k = signal.lfilter(rf_coeff, 1.0, \
				iq_data[(block_count)*block_size+1:(block_count+1)*block_size:2],
				zi=state_q_lpf_100k)

		# downsample the I/Q data from the FM channel
		i_ds = i_filt[::rf_decim]
		q_ds = q_filt[::rf_decim]

		# you MUST have also "custom" state-saving for your own FM demodulator
		dummy_fm, state_phase = fmDemodArctan(i_ds, q_ds, prev_phase = state_phase)
		#dummy_fm, dummy_state = myDemod(i_ds, q_ds, dummy_state)

		#print(len(i_ds))

    #RF front end stops here==================================================

		# GET RDS CHANNEL
		channel_filt, channel_block = block_convolution(channel_coeff, dummy_fm, channel_block)

		# LOWER PATH OF RDS CARRIER RECOVERY: MATCH DELAY OF BPF (USED IN MIXER)
		channel_Delay, delay_block = myAllPass(channel_filt, delay_block)

		# UPPER PATH OF RDS CARRIER RECOVERY:
		carrier_Input = sq_nonlinearity(channel_filt)
		carrier_filt, carrier_block = block_convolution(carrier_coeff, carrier_Input, carrier_block)
		recoveredRDSI, recoveredRDSQ, pll_block = fmPll(carrier_filt, 114e3, rf_Fs/rf_decim, 0.5, 1.1775, 0.003, pll_block)

        # RDS DEMODULATION
		mixedAudioI = mixer(recoveredRDSI, channel_Delay)
		mixedAudioQ = mixer(recoveredRDSQ, channel_Delay)

		demod_filtI, rds_demod_blockI = rs_block_convolution(rds_demod_coeff, mixedAudioI, rds_demod_blockI, demod_decim, demod_exp)
		demod_filtQ, rds_demod_blockQ = rs_block_convolution(rds_demod_coeff, mixedAudioQ, rds_demod_blockQ, demod_decim, demod_exp)

		demod_filtI, rds_rrc_blockI = block_convolution(rrc_coeff, demod_filtI, rds_rrc_blockI)
		demod_filtQ, rds_rrc_blockQ = block_convolution(rrc_coeff, demod_filtQ, rds_rrc_blockQ)

		# DO REST OF RDS AFTER BLOCK 1 (SINCE BLOCK 1 JUST FOR SET UP)
		if block_count >= 1:
			samplesQ, manchester_values, initialQ = CDR_state(demod_filtQ, sps, initialQ)
			samplesI, manchester_values, initialI = CDR_state(demod_filtI, sps, initialI)
			bits, decode_init = diff_decoding(manchester_values, decode_init, cdr_stateI)
			cdr_stateI = [manchester_values[-1], samplesI[-1], bits[-1]]

			#print("Last value: " + str(samplesI[-1]))
		#Generate Plots of Monopath
		if block_count >= 6 and block_count < 9:

			# IQ CONSTELLATION PLOTS
			in_phase = []
			quad = []

			for i in range(len(demod_filtI)):
				in_phase.append(demod_filtI[samplesI])

			for i in range(len(demod_filtQ)):
				quad.append(demod_filtQ[samplesQ])

			fig, ax = plt.subplots(1)
			fig.suptitle('constellation graphs')

			ax.set_xlabel("in_phase")
			ax.set_ylabel("quadrature")

			ax.scatter(in_phase, quad, s=10)
			ax.set_xlim([-1,1])
			ax.set_ylim([-1,1])

			n = 100
			x1 = range(n)
			x2 = range(n,n+n)

			fig2, axs = plt.subplots(4)

			axs[0].plot(x2, carrier_filt[:n], c='blue')
			axs[0].plot(x1, prev2[(len(prev2)-n):], c='orange')
			axs[0].set_title('Mixed Audio', fontstyle='italic',fontsize='medium')
			axs[0].axhline(y = 0, color = 'r', linestyle = '-')

			axs[1].plot(x2, recoveredRDSI[:n], c='blue')
			axs[1].plot(x1, prevPLLI[(len(prevPLLI)) - n:], c='orange')
			axs[1].set_title('PLL', fontstyle='italic',fontsize='medium')
			axs[1].axhline(y = 0, color = 'r', linestyle = '-')

			axs[2].plot(x2, recoveredRDSQ[:n], c='blue')
			axs[2].plot(x1, prevPLLQ[(len(prevPLLQ)) - n:], c='orange')
			axs[2].set_title('PLL', fontstyle='italic',fontsize='medium')
			axs[2].axhline(y = 0, color = 'r', linestyle = '-')

			axs[3].plot(range(len(demod_filtI)), demod_filtI, c='blue')
			axs[3].plot(samplesI, np.zeros(shape = len(samplesI)), marker="o")
			print("Length " + str(len(demod_filtI)))
			plt.show()

		#prev1 = demod_filt
		prev2 = carrier_filt
		prevPLLI = recoveredRDSI
		prevPLLQ = recoveredRDSQ
		print("===================================================================")
		"""
		if block_count >= 10 and block_count < 12:

			# plot PSD of selected block after FM demodulation
			ax0.clear()
			fmPlotPSD(ax0, dummy_fm, (rf_Fs/rf_decim)/1e3, subfig_height[0],
					'Demodulated FM (block ' + str(block_count) + ')')
			# output binary file name (where samples are written from Python)
			fm_demod_fname = "../data/fm_demod_" + str(block_count) + ".bin"
			# create binary file where each sample is a 32-bit float
			dummy_fm.astype('float32').tofile(fm_demod_fname)

			# plot PSD of selected block after extracting mono audio
			# ... change as needed
			fmPlotPSD(ax1, audio_block, ((rf_Fs/rf_decim)/1e3), subfig_height[1], 'mixed audio')

			fmPlotPSD(ax2, right_data, ((rf_Fs/rf_decim)/1e3), subfig_height[2], 'Right Channel')
			# plot PSD of selected block after downsampling mono audio
			# ... change as needed
			fmPlotPSD(ax3, left_data, audio_Fs/1e3, subfig_height[3], 'Left Channel')
			# save figure to file
			fig.savefig("../data/fmMonoBlockDelay" + str(block_count) + ".png")
		"""
		block_count += 1

	print('Finished processing all the blocks from the recorded I/Q samples')

	complete_data = np.vstack((left_data, right_data)).T

	# write audio data to file
	out_fname = "../data/fmStereoOldDemod.wav"
	wavfile.write(out_fname, int(audio_Fs), np.int16((complete_data/2)*4095))
	print("Written audio samples to \"" + out_fname + "\" in signed 16-bit format")

	# uncomment assuming you wish to show some plots
	# plt.show()

File: src/fmRDSDesign.cpp
#include "dy4.h"
#include "filter.h"
#include "fourier.h"
#include "genfunc.h"
#include "iofunc.h"
#include "logfunc.h"
#include "block_conv_fn.h"
#include "mono_path.h"

#include <stdio.h>
#include <math.h>

// testing time complexity
#include <chrono>

void readStdinBlockData(unsigned int num_samples, std::vector<float> &block_data){
  std::vector<char> raw_data(num_samples);
  std::cin.read(reinterpret_cast<char*>(&raw_data[0]), num_samples*sizeof(char));
  for(int k = 0; k < (int)num_samples; k++) {
    block_data[k] = float(((unsigned char)raw_data[k] - 128)/ 128.0);
  }
}

std::vector<float> rds_path_main(int mode){
	// TIMING VARIABLES
  auto start_overall_time = std::chrono::high_resolution_clock::now();
  auto start_time = std::chrono::high_resolution_clock::now();
  auto stop_time = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double, std::milli> RDS_run_time;
  std::chrono::duration<double, std::milli> RF_run_time;

  // Determine custom parameters depending on mode
  float rf_Fs;
  int rf_decim, sps, demod_decim, demod_exp;

  switch(mode) {
    case 0:
      rf_Fs = 2400000.0;
      rf_decim = 10;
      sps = 17;
      demod_decim = 1920;
      demod_exp = 323;
      break;
    case 1:
      rf_Fs = 1440000.0;
      rf_decim = 5;
      // HANDLE WITH PRINT STATEMENT
      break;
    case 2:
      rf_Fs = 2400000.0;
      rf_decim = 10;
      sps = 30;
      demod_decim = 1; // NOT CALCULATED
      demod_exp = 1; // NOT CALCULATED
      break;
    case 3:
      rf_Fs = 2304000.0;
      rf_decim = 9;
      // HANDLE WITH PRINT STATEMENT
      break;
  }

  // RF front end variables
	float rf_Fc = 100000.0;
	int rf_taps = 151;
  int audio_taps = 151;
  // RDS path variables
  int rds_sr = sps * 2375;

  // filter coefficients for RDS
  std::vector<float> rf_coeff, carrier_coeff, channel_coeff, rds_demod_coeff, rds_rrc_coeff;
	low_pass_coeff(rf_Fs, rf_Fc, rf_taps, rf_coeff);
	band_pass_coeff(113500, 114500, rf_Fs/rf_decim, audio_taps, carrier_coeff);
	band_pass_coeff(54000, 60000, rf_Fs/rf_decim, audio_taps, channel_coeff);
	low_pass_coeff((rf_Fs/rf_decim)*demod_exp, 3000, audio_taps*demod_exp, rds_demod_coeff);
  rrc_coeff(rds_sr, rf_taps, rds_rrc_coeff);
	int block_size = 38400*2;

  // regular IQ data
  std::vector<float> i_data, q_data;
  i_data.resize(block_size / 2);
  q_data.resize(block_size / 2);

  // filtered IQ data
  std::vector<float> i_filt, q_filt;
  i_filt.resize(i_data.size());
  q_filt.resize(q_data.size());

  // downsampled filtered IQ data
  std::vector<float>i_ds, q_ds;
  i_ds.resize(i_filt.size()/rf_decim);
  q_ds.resize(q_filt.size()/rf_decim);

  // state saving variables for I and Q samples convolution
  std::vector<float> state_i_lpf_100k, state_q_lpf_100k;
  state_i_lpf_100k.resize(rf_coeff.size() - 1, 0.0);
  state_q_lpf_100k.resize(rf_coeff.size() - 1, 0.0);

  // demodulation variables
	std::vector<float> demod_state, IQ_demod;
	demod_state.resize(2, 0.0);

  // RDS filtered data
  std::vector<float> carrier_filt, channel_filt, recoveredRDS, rds_demod_filt, rrc_demod_filt, demod_down, rds_demod_block;
  carrier_filt.resize(i_data.size() - 1);
  channel_filt.resize(i_data.size() - 1);
  rds_demod_filt.resize(demod_exp*(i_data.size() - 1));
  rds_demod_block.resize(rds_demod_filt.size()/demod_decim);

  rrc_demod_filt.resize(rds_demod_block.size());


  // state saving variables for RDS path
	std::vector<float> carrier_state, channel_state, rds_demod_state, rds_rrc_state, pll_state;
	carrier_state.resize(carrier_coeff.size() - 1);	// for RDS Carrier Recovery
	channel_state.resize(channel_coeff.size() - 1);	// for RDS Channel Extraction
  rds_demod_state.resize(rds_demod_coeff.size() - 1); // for RDS Rational Resampler
  rds_rrc_state.resize(rds_rrc_coeff.size() - 1); // for RDS RRC convolution

  pll_state.resize(6);
  pll_state[0] = 0.0;
  pll_state[1] = 0.0;
  pll_state[2] = 1.0;
  pll_state[3] = 0.0;
  pll_state[4] = 1.0;
  pll_state[5] = 0.0;

  // RDS demodulation
  std::vector<float> mixed_audio, delay_block, channel_delay;
  mixed_audio.resize(i_data.size() - 1);
  delay_block.resize((rf_taps - 1)/2);

  float prev_phase = 0.0;

  int cdr_init = -1;
  int decode_init = -1;
  std::vector<int> cdr_state = {0, 0, 0};
  std::vector<int> sample_idxs, man_encoding, decoded_bits;
  int last_bit;
  // decipher each block
	for(unsigned int block_id = 0; ; block_id++) {
		std::vector<float> block_data(block_size);
    readStdinBlockData(block_size, block_data);
    if ((std::cin.rdstate()) != 0) {
      std::cerr << "End of input stream reached" << std::endl;

      // timing analysis
      stop_time = std::chrono::high_resolution_clock::now();
		  std::chrono::duration<double, std::milli> OVERALL_run_time = stop_time-start_overall_time;
		  std::cerr << "OVERALL RUNTIME: " << OVERALL_run_time.count() << " ms" << "\n";
      std::cerr << "RF DOWNSAMPLE RUNTIME: " << RF_run_time.count() << " ms" << "\n";
      std::cerr << "TOTAL RDSPATH RUNTIME: " << RDS_run_time.count() << " ms" << "\n";
      exit(1);
    }

    //std::cerr << "Processing Block " << block_id << std::endl;

    // STEP 1: IQ samples demodulation
    // take IQ data in
		for (int k = 0; k < block_size / 2; k++) {
      i_data[k] = block_data[2*k];
      q_data[k] = block_data[2*k + 1];
    }

    // filter out IQ data with convolution
    start_time = std::chrono::high_resolution_clock::now();
    i_ds.clear();
    q_ds.clear();
		ds_block_conv(i_filt,i_data,rf_coeff,state_i_lpf_100k,rf_decim,i_ds);
		ds_block_conv(q_filt,q_data,rf_coeff,state_q_lpf_100k,rf_decim,q_ds);

    // perform demodulation on IQ data
    IQ_demod = fmDemodArctan(i_ds, q_ds, prev_phase);

    // timing analysis
    stop_time = std::chrono::high_resolution_clock::now();
    RF_run_time += stop_time-start_time;

		// STEP 2: RDS path
    start_time = std::chrono::high_resolution_clock::now();

    //RDS Channel Extraction: all-pass
		state_block_conv(channel_filt, IQ_demod, channel_coeff, channel_state);
    all_pass_coeff(channel_delay, channel_filt, delay_block);

		//RDS Carrier Recovery: non-linearity -> Bandpass -> PLL -> Numerically Controlled
    std::vector<float> carrier_input = sq_non_linearity(channel_filt);

		state_block_conv(carrier_filt, carrier_input, carrier_coeff, carrier_state);

		fmPll(carrier_filt, recoveredRDS, pll_state, 114e3, rf_Fs/rf_decim, 0.5, 1.1775, 0.003);

    //RDS Demodulation: mixer -> digital filtering (lowpass -> resampler) -> rrc -> CDR
		mixer(recoveredRDS, channel_delay, mixed_audio);

    rds_demod_block.clear();
    rs_block_conv(rds_demod_filt, mixed_audio, rds_demod_coeff, rds_demod_state, demod_decim, demod_exp, rds_demod_block);
    state_block_conv(rrc_demod_filt, rds_demod_block, rds_rrc_coeff, rds_rrc_state);

    if (block_id >= 1){
      CDR(rrc_demod_filt, sps, cdr_init, sample_idxs, man_encoding);
      last_bit = diff_decoding(man_encoding, cdr_state, decoded_bits, decode_init);
      printRealVector(decoded_bits);
      cdr_state = {man_encoding[man_encoding.size() - 1], sample_idxs[sample_idxs.size() - 1], last_bit};
    }
    //...

    // timing analysis
    stop_time = std::chrono::high_resolution_clock::now();
    RDS_run_time += stop_time-start_time;
  }
}

int main(int argc, char* argv[])
{
  int mode = 0;

  // default operation
  if (argc < 2){
    std::cerr << "Operating in default mode 0" << std::endl;
  }
  // mode operation
  else if (argc == 2){
    mode = atoi(argv[1]);
    if (mode > 3){
      std::cerr << "Not a valid mode: " << mode << std::endl;
      exit(1);
    }
  }
  else{
    std::cerr << "Usage: " << argv[0] << std::endl;
    std::cerr << "or " << std::endl;
    std::cerr << "Usage: " << argv[0] << " <mode>" << std::endl;
    std::cerr << "\t\t <mode> is a value from 0 to 3" << std::endl;
    exit(1);
  }

  std::cerr << "Operating in mode " << mode << std::endl;
  rds_path_main(mode);

	return 0;
}
